### PR TITLE
Added Support for Static Assignments

### LINKED ISSUE #27: Support Static Assignments, Per-host Options
It would be nice to support static assignment, and sending specific options (and values) to each host. For example, only responding to one specific host is desirable, rather than any PXE DHCP request received.

### PR DESCRIPTION
**Pull request by psychomario**
This is just minimal support. It probably needs extending for wildcards and ranges.

Also note we were never actually sending DNS servers as far as I can tell. This has now been added. 

This closes #27.

### CODE REVIEW

File: pypxe-server.py [hunk 1]
```diff
@@ -137,7 +153,8 @@
                     useipxe = args.USE_IPXE,
                     usehttp = args.USE_HTTP,
                     mode_proxy = args.DHCP_MODE_PROXY,
-                    mode_debug = ("dhcp" in args.MODE_DEBUG.lower() or "all" in args.MODE_DEBUG.lower()))
+                    mode_debug = ("dhcp" in args.MODE_DEBUG.lower() or "all" in args.MODE_DEBUG.lower()),
```
Comment (mmattioli) [inline pypxe-server.py]: why are you checking for "all"?

File: example.json [hunk 2]
```diff
@@ -13,7 +13,7 @@
 "USE_IPXE":false,
 "USE_HTTP":false,
 "USE_TFTP":true,
-"MODE_DEBUG":false,
+"MODE_DEBUG":"",
```
Comment (mmattioli) [inline example.json]: why are you changing it to ""?

File: pypxe-server.py (hunk 1 above)
Comment (psychomario) [inline pypxe-server.py]: That was there before this PR, but allows you to say `--debug all` vs. `--debug tftp,http,dhcp`

Comment (mmattioli) [inline pypxe-server.py]: I think you need to rebase `staticassignments` off of `development` :)

**Comment by mmattioli** [comment]
Has any of this been tested?

File: example.json (hunk 2 above)
Comment (psychomario) [inline example.json]: `"dhcp" in args.MODE_DEBUG.lower()` for example won't work before as it's a bool. This value should be a string now.

**Comment by psychomario** [comment]
50%. I'm having a crashing issue which I need to look into first, as it's crashing half way through this part. See [this](https://github.com/psychomario/PyPXE/pull/59#issuecomment-93786512) comment for details.

File: pypxe-server.py (hunk 1 above)
Comment (psychomario) [inline pypxe-server.py]: Unless I am completely mistaken as to what rebase does, the reason this line is highlighted is becuase I'm adding an argument after it, changing the last parenthesis to a comma.

Comment (mmattioli) [inline pypxe-server.py]: Apologies, I had a brain fart there for a second. You're right, I just wasn't thinking clearly as to what you were looking for here. Never mind :)

**Comment by mmattioli** [comment]
As per our [conversation](https://github.com/psychomario/PyPXE/pull/59#issuecomment-94306273), this will not be merged into `development` until #59 is merged and `development` is merged into `master`. Please keep in mind it may need to be rebased once #59 is merged.

**Comment by psychomario** [comment]
Rebase complete. This PR now also explicitly fixes #79 and implicitly fixes #77.

Support for wildcards, ranges and filenames: I will try to get to this over the next week or so.

@mmattioli Regarding wildcards/ranges, are you against regular expressions or do you think rolling our own matching would be preferrable?

**Comment by psychomario** [comment]
Note, this is not ready yet. Upon re-reading #27, a whitelist was suggested. This shouldn't be too hard to implement but until such time, this PR is not ready to be merged.

**Comment by psychomario** [comment]
Whitelist has now been implemented, along with missing documentation in the README. This is now ready to be merged, pending testing for a bug I've just realised I might have introduced.

**Comment by psychomario** [comment]
This has now been tested. I fixed a few bugs and oddly there doesn't seem to be the same unicode issue seen before. The whitelist makes it very easy to test DHCP server changes on live networks now, added bonus.

**Comment by mmattioli** [comment]
@psychomario is this tested and ready for merge? I want to merge this before #70 (which will require #70 to be rebased but shouldn't be too big of a hassle) but I don't want to merge _anything_ into `development` until #78 is tested and complete for obvious reasons.

**Comment by psychomario** [comment]
Yes this PR is ready to go.

There's no way of getting out of rebase conflicts so there's no benefit of doing either first in that respect, but #70 first sounds good.

I didn't expect this to be merged before #78, it just gave me something to do until I can do DHCP testing tomorrow.

**Comment by mmattioli** [comment]
No problem, just wanted to confirm the timeline with you :+1:

**Comment by mmattioli** [comment]
@psychomario let me know when this is ready to merge after you've rebased

**Comment by psychomario** [comment]
I think this is ready now, unless you want #84 first.

**Comment by mmattioli** [comment]
@psychomario, your call; whichever one that gets merged first will require a rebase of the other. Which would you rather merge first?

**Comment by psychomario** [comment]
I think it would be easiest to merge this one first. I'm not 100% but #84 might actually rebase cleanly on top of this one.

**Comment by psychomario** [comment]
This PR closes #79

**Comment by mmattioli** [comment]
>@mmattioli Regarding wildcards/ranges, are you against regular expressions or do you think rolling our own matching would be preferrable?

@psychomario I think regular expressions should be avoided at all costs regardless of the project and/or implementation

**Comment by psychomario** [comment]
Thinking on this I've actually realised that wildcards and ranges don't apply to leases, only prefixes, so regular expressions isn't even a (bad)  solution.

I'll do some research to find what kind of configurations are used and then write something to solve it.

**Pull request merged**